ckan/lib/uploader.py

```diff
@@ -404,3 +412,264 @@ def upload(self, id: str, max_size: int = 10) -> None:
+                }
+            )
+
+        # Check that the declared types in the request are supported
+        declared_mimetype_from_filename = mimetypes.guess_type(
+            self.upload_field_storage.filename
+        )[0]
+        declared_content_type = self.upload_field_storage.content_type
+        for declared_mimetype in (
+            declared_mimetype_from_filename,
+            declared_content_type,
+        ):
+            if (
+                declared_mimetype
+                and allowed_mimetypes
+                and allowed_mimetypes[0] != "*"
+                and declared_mimetype not in allowed_mimetypes
+            ):
+                raise logic.ValidationError(
+                    {
+                        self.file_field: [
+                            f"Unsupported upload type: {declared_mimetype}"
+                        ]
+                    }
+                )
+
+        # Check that the actual type guessed from the contents is supported
+        # (2KB required for detecting xlsx mimetype)
+        guessed_mimetype = self.upload_file.content_type
+
+        err: ErrorDict = {
+            self.file_field: [f"Unsupported upload type: {guessed_mimetype}"]
+        }
+
+        if (allowed_mimetypes
+                and allowed_mimetypes[0] != "*"
+                and guessed_mimetype not in allowed_mimetypes):
+            raise logic.ValidationError(err)
+
+        type_ = guessed_mimetype.split("/")[0]
+        if allowed_types and allowed_types[0] != "*" and type_ not in allowed_types:
+            raise logic.ValidationError(err)
+
+        preferred_extension = mimetypes.guess_extension(guessed_mimetype)
+        if preferred_extension and self.filename:
+            self.filename = str(Path(self.filename).with_suffix(preferred_extension))
+
+
+class FkResourceUpload(object):
+    mimetype: Optional[str]
+    storage: fk.Storage | None = None
+
+    def __init__(self, resource: dict[str, Any]) -> None:
+        try:
+            self.storage = files.get_storage(
+                config["ckan.files.default_storages.resource"]
+            )
+        except files.exc.UnknownStorageError:
+            log.warning("Resource storage is not configured")
+            return
+
+        config_mimetype_guess = config.get('ckan.mimetype_guess')
+
+        self.filename = None
+        self.mimetype = None
+
+        upload_field_storage = resource.pop('upload', None)
+        self.clear = resource.pop('clear_upload', None)
+
+        if bool(upload_field_storage) and \
+                isinstance(upload_field_storage, ALLOWED_UPLOAD_TYPES):
+
+            self.upload_file = files.make_upload(upload_field_storage)
+            self.filesize = self.upload_file.size
+            self.filename = munge.munge_filename(self.upload_file.filename)
+            resource['url'] = self.filename
+            resource['url_type'] = 'upload'
+            resource['last_modified'] = datetime.datetime.utcnow()
```

Comment:
As above, I'd really prefer we had a clearer, simpler interface if possible.

It's very surprising to have the `resource` dict parameter values modified after constructing a `FkResourceUpload` instance.